controller/src/dns/mod.rs

```diff
@@ -0,0 +1,218 @@
+            //     let name = request.query().name();
+            //     let rdata = RData::SOA(SOA::new(
+            //         Name::from_ascii(&self.spawner.nameserver)
+            //             .or_dns_error(ResponseCode::ServFail, || {
+            //                 format!("Failed to encode nameserver {}.", self.spawner.nameserver)
+            //             })?,
+            //         Name::from_ascii("paul.driftingin.space")
+            //             .or_dns_error(ResponseCode::ServFail, || {
+            //                 format!("Failed to encode nameserver contact.")
+            //             })?,
+            //         1,
+            //         7200,
+            //         7200,
+            //         7200,
+            //         7200,
+            //     ));
+            //     let ttl = 60;
+            //     let record = Record::from_rdata(name.clone().into(), ttl, rdata);
+
+            //     Ok(vec![record])
+            // }
+            RecordType::CAA | RecordType::AAAA => Ok(vec![]), // Not supported but don't report.
+            request => {
+                tracing::info!(?request, "Unhandled lookup type {}.", request);
+                Ok(vec![])
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl RequestHandler for ClusterDnsServer {
+    async fn handle_request<R>(&self, request: &Request, mut response_handle: R) -> ResponseInfo
+    where
+        R: ResponseHandler,
+    {
+        let builder = MessageResponseBuilder::from_message_request(request);
+        let mut header = Header::response_from_request(request.header());
+
+        let result = match self.do_lookup(request).await {
+            Ok(answers) => {
+                let response = builder.build(header, answers.iter(), vec![], vec![], vec![]);
+                response_handle.send_response(response).await
+            }
+            Err(e) => {
+                tracing::warn!(?e, ?request, "Error encountered processing DNS request.");
+
+                header.set_response_code(e.code);
+                let response = builder.build_no_records(header);
+                response_handle.send_response(response).await
+            }
+        };
+
+        if let Ok(result) = result {
+            result
+        } else {
+            tracing::warn!(?request, "send_response failed in DNS handling.");
+            ResponseInfo::from(header)
+        }
+    }
+}
+
+pub async fn serve_dns(plan: DnsPlan) -> anyhow::Result<Never> {
+    let mut fut = ServerFuture::new(ClusterDnsServer::new(&plan.nc).await);
+
+    let ip_port_pair = (plan.options.bind_ip, plan.options.port);
+
+    let sock = UdpSocket::bind(ip_port_pair)
+        .await
+        .context("Binding UDP port for DNS server.")?;
+    fut.register_socket(sock);
+
+    let listener = TcpListener::bind(ip_port_pair)
+        .await
+        .context("Binding TCP port for DNS server.")?;
+    fut.register_listener(listener, Duration::from_secs(10));
```

Comment:
the 10 here should be in a more descriptive const (something like MAX_TCP_CONNECTION_DURATION_SECS)

Reply:
Fixed.